modules/exadsBidAdapter.js

```diff
@@ -0,0 +1,433 @@
+
+    const bidRq = JSON.parse(request.data);
+    const requestId = serverResponse.body.id;
+    const bidData = serverResponse.body.seatbid[0].bid[0];
+    const currency = serverResponse.body.cur;
+    const bidResponseAd = bidData.adm;
+    const pixelUrl = bidData.nurl.replace(/^http:\/\//i, 'https://');
+
+    const bannerInfo = utils.deepAccess(bidRq.imp[0], 'banner');
+    const nativeInfo = utils.deepAccess(bidRq.imp[0], 'native');
+    const videoInfo = utils.deepAccess(bidRq.imp[0], 'video');
+
+    let w; let h = 0;
+    let mediaType = '';
+    const native = {};
+
+    if (bannerInfo != null) {
+      w = bidRq.imp[0].banner.w;
+      h = bidRq.imp[0].banner.h;
+      mediaType = BANNER;
+    } else if (nativeInfo != null) {
+      const reqNative = JSON.parse(bidResponseAd);
+      reqNative.native.assets.forEach(asset => {
+        if (asset.img != null) {
+          const imgAsset = JSON.parse(bidRq.imp[0].native.request)
+            .native.assets.filter(asset => asset.img != null).map(asset => asset.img);
+          w = imgAsset[0].w;
+          h = imgAsset[0].h;
+          native.image = {
+            url: asset.img.url,
+            height: h,
+            width: w
+          }
+        } else if (asset.title != null) {
+          native.title = asset.title.text;
+        } else if (asset.data != null) {
+          native.body = asset.data.value;
+        } else {
+          utils.logWarn('bidResponse->', 'wrong asset type or null');
+        }
+      });
+      mediaType = NATIVE;
+    } else if (videoInfo != null) {
+      mediaType = VIDEO;
+    }
+
+    const bidResponse = {
+      requestId: requestId,
+      currency: currency,
+      ad: bidData.adm,
+      cpm: bidData.price,
+      creativeId: bidData.crid,
+      cid: bidData.cid,
+      width: w,
+      ttl: 360,
+      height: h,
+      netRevenue: true,
+      mediaType: mediaType,
+      nurl: pixelUrl
+    };
+
+    if (mediaType == 'native') {
+      native.clickUrl = bidData.adomain[0];
+      bidResponse.native = native;
+    }
+
+    if (mediaType == 'video') {
+      bidResponse.vastXml = bidData.adm;
+      bidResponse.width = bidData.w;
+      bidResponse.height = bidData.h;
+    }
+
+    utils.logInfo('bidResponse->', bidResponse);
+
+    bidResponses.push(bidResponse);
+  } else {
+    utils.logInfo('NO Ad server response ->', serverResponse.body.id);
+  }
+
+  utils.logInfo('interpretResponse -> bidResponses:', bidResponses);
+
+  return bidResponses;
+}
+
+function makeBidRequest(url, data) {
+  const payloadString = JSON.stringify(data);
+
+  return {
+    method: 'POST',
+    url: url,
+    data: payloadString,
+  }
+}
+
+function getUrl(adPartner, bid) {
+  let endpointUrlMapping = {
+    [PARTNERS.RTB_2_4]: bid.params.endpoint + '?idzone=' + bid.params.zoneId + '&fid=' + bid.params.fid
+  };
+
+  return endpointUrlMapping[adPartner] ? endpointUrlMapping[adPartner] : 'defaultEndpoint';
+}
+
+function manageEnvParams() {
```

Comment:
I'm not a big fan of using a "global" `envParams` object. Given that we only use it for the request, why not creating an helper function to return the params we need, something like ` fn get_env_params() -> env_params`.